File: src/oauth/authcode.ts
// PKCE authorization-code flow over a loopback redirect (RFC 8252).
//
// Runs the OAuth authorize step that yields an authorization `code`; the DPoP
// token exchange is P04. There is NO DPoP at the authorize step. The redirect
// lands on an ephemeral HTTP server bound to 127.0.0.1 only. Never log the
// code, verifier, or state.
import { createHash, randomBytes } from "node:crypto";
import type { Config } from "../config.ts";
import { logger as defaultLogger, type Logger } from "../logger.ts";
import type { Endpoints } from "./discovery.ts";

/** base64url(bytes) with no padding. */
function base64url(bytes: Buffer): string {
  return bytes.toString("base64url");
}

export interface Pkce {
  verifier: string;
  challenge: string;
}

/**
 * Generate a PKCE verifier/challenge pair. The verifier is a 43-char base64url
 * string (32 random bytes), within the RFC 7636 [43,128] range; the challenge
 * is base64url(sha256(verifier)) (S256).
 */
export function generatePkce(): Pkce {
  const verifier = base64url(randomBytes(32));
  const challenge = base64url(createHash("sha256").update(verifier).digest());
  return { verifier, challenge };
}

/** Open a URL in the user's default browser. */
export type Opener = (url: string) => void | Promise<void>;

function defaultOpener(url: string): void {
  const cmd =
    process.platform === "darwin"
      ? ["open", url]
      : process.platform === "win32"
        ? ["cmd", "/c", "start", "", url]
        : ["xdg-open", url];
  Bun.spawn(cmd, { stdout: "ignore", stderr: "ignore", stdin: "ignore" });
}

export interface AuthorizeOptions {
  /** Injectable browser opener; defaults to the OS opener. */
  opener?: Opener;
  /** Reject after this many ms (default 300_000). */
  timeoutMs?: number;
  logger?: Logger;
}

export interface AuthorizeResult {
  code: string;
  redirectUri: string;
  verifier: string;
}

const CLOSE_PAGE = (message: string): string =>
  `<!doctype html><html><head><meta charset="utf-8"><title>okta-mcp-bridge</title></head>` +
  `<body style="font-family:system-ui;padding:2rem"><p>${message}</p>` +
  `<p>You may close this tab and return to your terminal.</p></body></html>`;

/**
 * Run the PKCE authorization-code flow. Starts a loopback server, opens the
 * authorize URL in the browser, and resolves with the captured code once the
 * redirect arrives. Rejects on state mismatch, an `error` param, or timeout.
 */
export function authorize(
  config: Config,
  endpoints: Endpoints,
  opts: AuthorizeOptions = {},
): Promise<AuthorizeResult> {
  const logger = opts.logger ?? defaultLogger;
  const opener = opts.opener ?? defaultOpener;
  const timeoutMs = opts.timeoutMs ?? 300_000;

  const { verifier, challenge } = generatePkce();
  const state = base64url(randomBytes(32));

  return new Promise<AuthorizeResult>((resolve, reject) => {
    let settled = false;
    let timer: ReturnType<typeof setTimeout> | undefined;

    const server = Bun.serve({
      hostname: "127.0.0.1",
      port: config.OKTA_REDIRECT_PORT,
      fetch(req): Response {
        const url = new URL(req.url);
        if (url.pathname !== "/callback") {
          return new Response("Not found", { status: 404 });
        }
        const htmlHeaders = { "Content-Type": "text/html; charset=utf-8" };

        const errParam = url.searchParams.get("error");
        if (errParam) {
          const desc = url.searchParams.get("error_description") ?? "";
          settle(() =>
            reject(
              new Error(`authorization failed: ${errParam}${desc ? ` — ${desc}` : ""}`),
            ),
          );
          return new Response(CLOSE_PAGE("Authorization failed."), {
            status: 400,
            headers: htmlHeaders,
          });
        }

        if (url.searchParams.get("state") !== state) {
          settle(() => reject(new Error("authorization failed: state mismatch")));
          return new Response(CLOSE_PAGE("Authorization failed (state mismatch)."), {
            status: 400,
            headers: htmlHeaders,
          });
        }

        const code = url.searchParams.get("code");
        if (!code) {
          settle(() => reject(new Error("authorization failed: no code in callback")));
          return new Response(CLOSE_PAGE("Authorization failed (no code)."), {
            status: 400,
            headers: htmlHeaders,
          });
        }

        logger.info("oauth.authorize.code_received");
        settle(() => resolve({ code, redirectUri, verifier }));
        return new Response(CLOSE_PAGE("Authorization complete."), {
          headers: htmlHeaders,
        });
      },
    });

    const redirectUri = `http://127.0.0.1:${server.port}/callback`;

    function settle(action: () => void): void {
      if (settled) return;
      settled = true;
      if (timer) clearTimeout(timer);
      action();
      // Let the in-flight response flush before stopping the server.
      setTimeout(() => server.stop(), 0);
    }

    timer = setTimeout(() => {
      logger.warn("oauth.authorize.failed", { reason: "timeout" });
      settle(() => reject(new Error("authorization timed out")));
    }, timeoutMs);

    const authUrl = new URL(endpoints.authorizationEndpoint);
    authUrl.searchParams.set("response_type", "code");
    authUrl.searchParams.set("client_id", config.OKTA_CLIENT_ID);
    authUrl.searchParams.set("redirect_uri", redirectUri);
    authUrl.searchParams.set("scope", config.OKTA_SCOPES);
    authUrl.searchParams.set("state", state);
    authUrl.searchParams.set("code_challenge", challenge);
    authUrl.searchParams.set("code_challenge_method", "S256");

    logger.info("oauth.authorize.started", { redirect_port: server.port });

    Promise.resolve()
      .then(() => opener(authUrl.toString()))
      .catch((cause) => {
        // The opener is best-effort: fall back to manual instructions on stderr.
        logger.warn("oauth.authorize.opener_failed", { error: String(cause) });
        process.stderr.write(
          `\nOpen this URL in your browser to authorize:\n\n${authUrl.toString()}\n\n`,
        );
      });
  });
}

File: internal/oauth/authcode.go
package oauth

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"runtime"
	"sync"
	"time"

	"github.com/joewitt99/bridge-mcp-client/internal/config"
	"github.com/joewitt99/bridge-mcp-client/internal/logx"
)

// AuthCodeResult is the output of the authorize step (consumed by ExchangeCode).
type AuthCodeResult struct {
	Code        string
	RedirectURI string
	Verifier    string
}

// Opener opens a URL in the user's browser (injectable for tests).
type Opener func(url string) error

// GeneratePKCE returns a PKCE verifier (43-char base64url) and its S256
// challenge = base64url(sha256(verifier)).
func GeneratePKCE() (verifier, challenge string, err error) {
	b := make([]byte, 32)
	if _, err = rand.Read(b); err != nil {
		return "", "", err
	}
	verifier = base64.RawURLEncoding.EncodeToString(b)
	sum := sha256.Sum256([]byte(verifier))
	challenge = base64.RawURLEncoding.EncodeToString(sum[:])
	return verifier, challenge, nil
}

// AuthorizeOptions configure Authorize (all optional).
type AuthorizeOptions struct {
	Opener  Opener
	Timeout time.Duration
	Logger  *logx.Logger
}

const closePage = `<!doctype html><html><head><meta charset="utf-8"><title>okta-mcp-bridge</title></head>` +
	`<body style="font-family:system-ui;padding:2rem"><p>%s</p>` +
	`<p>You may close this tab and return to your terminal.</p></body></html>`

// Authorize runs the PKCE authorization-code flow over a 127.0.0.1 loopback
// redirect and returns the captured code. There is NO DPoP at this step.
func Authorize(cfg config.Config, ep Endpoints, opts AuthorizeOptions) (AuthCodeResult, error) {
	logger := opts.Logger
	if logger == nil {
		logger = logx.Default
	}
	timeout := opts.Timeout
	if timeout <= 0 {
		timeout = 300 * time.Second
	}
	opener := opts.Opener
	if opener == nil {
		opener = defaultOpener
	}

	verifier, challenge, err := GeneratePKCE()
	if err != nil {
		return AuthCodeResult{}, err
	}
	state, err := randomState()
	if err != nil {
		return AuthCodeResult{}, err
	}

	if cfg.OktaRedirectPort == 0 {
		// Okta matches the redirect URI exactly, including the port, and does
		// not honor ephemeral ports — even with a wildcard registered.
		logger.Warn("oauth.authorize.ephemeral_port", logx.Fields{
			"note": "Okta requires a fixed, pre-registered redirect port; set OKTA_REDIRECT_PORT to match the redirect URI registered in Okta",
		})
	}

	ln, err := net.Listen("tcp", fmt.Sprintf("127.0.0.1:%d", cfg.OktaRedirectPort))
	if err != nil {
		return AuthCodeResult{}, fmt.Errorf("loopback listen: %w", err)
	}
	port := ln.Addr().(*net.TCPAddr).Port
	redirectURI := fmt.Sprintf("http://127.0.0.1:%d/callback", port)

	type outcome struct {
		code string
		err  error
	}
	resultCh := make(chan outcome, 1)
	var once sync.Once
	settle := func(o outcome) { once.Do(func() { resultCh <- o }) }

	mux := http.NewServeMux()
	mux.HandleFunc("/callback", func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		page := func(status int, msg string) {
			w.Header().Set("Content-Type", "text/html; charset=utf-8")
			w.WriteHeader(status)
			fmt.Fprintf(w, closePage, msg)
		}
		if e := q.Get("error"); e != "" {
			msg := e
			if d := q.Get("error_description"); d != "" {
				msg = e + " — " + d
			}
			page(http.StatusBadRequest, "Authorization failed.")
			settle(outcome{err: fmt.Errorf("authorization failed: %s", msg)})
			return
		}
		if q.Get("state") != state {
			page(http.StatusBadRequest, "Authorization failed (state mismatch).")
			settle(outcome{err: fmt.Errorf("authorization failed: state mismatch")})
			return
		}
		code := q.Get("code")
		if code == "" {
			page(http.StatusBadRequest, "Authorization failed (no code).")
			settle(outcome{err: fmt.Errorf("authorization failed: no code in callback")})
			return
		}
		logger.Info("oauth.authorize.code_received", nil)
		page(http.StatusOK, "Authorization complete.")
		settle(outcome{code: code})
	})

	srv := &http.Server{Handler: mux}
	go func() { _ = srv.Serve(ln) }()

	authURL := buildAuthorizeURL(cfg, ep, redirectURI, state, challenge)
	logger.Info("oauth.authorize.started", logx.Fields{"redirect_port": port})
	go func() {
		if err := opener(authURL); err != nil {
			logger.Warn("oauth.authorize.opener_failed", logx.Fields{"error": err.Error()})
			fmt.Fprintf(os.Stderr, "\nOpen this URL in your browser to authorize:\n\n%s\n\n", authURL)
		}
	}()

	timer := time.NewTimer(timeout)
	defer timer.Stop()
	var res outcome
	select {
	case res = <-resultCh:
	case <-timer.C:
		logger.Warn("oauth.authorize.failed", logx.Fields{"reason": "timeout"})
		res = outcome{err: fmt.Errorf("authorization timed out")}
	}

	shutdownCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	_ = srv.Shutdown(shutdownCtx)

	if res.err != nil {
		return AuthCodeResult{}, res.err
	}
	return AuthCodeResult{Code: res.code, RedirectURI: redirectURI, Verifier: verifier}, nil
}

func randomState() (string, error) {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(b), nil
}

func buildAuthorizeURL(cfg config.Config, ep Endpoints, redirectURI, state, challenge string) string {
	u, err := url.Parse(ep.AuthorizationEndpoint)
	if err != nil {
		// Fall back to a best-effort string; discovery already validated this.
		return ep.AuthorizationEndpoint
	}
	q := u.Query()
	q.Set("response_type", "code")
	q.Set("client_id", cfg.OktaClientID)
	q.Set("redirect_uri", redirectURI)
	q.Set("scope", cfg.OktaScopes)
	q.Set("state", state)
	q.Set("code_challenge", challenge)
	q.Set("code_challenge_method", "S256")
	u.RawQuery = q.Encode()
	return u.String()
}

func defaultOpener(rawURL string) error {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		cmd = exec.Command("open", rawURL)
	case "windows":
		// rundll32 avoids cmd.exe's `&`-in-URL pitfalls with OAuth query strings.
		cmd = exec.Command("rundll32", "url.dll,FileProtocolHandler", rawURL)
	default:
		cmd = exec.Command("xdg-open", rawURL)
	}
	return cmd.Start()
}
